### PR TITLE
Add a no-short-mac FIPS indicator

### PR DESCRIPTION
**Pull request by paulidale**
KMAC is not permitted to produce less than 32 bits of output as per SP 800-185 8.4.2.  This change enforces this.

HMAC is also subject to the same limitation but our implementation only outputs the digest block size.

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: include/openssl/fips_names.h [hunk 1]
```diff
@@ -61,6 +61,14 @@ extern "C" {
  */
 # define OSSL_PROV_FIPS_PARAM_TLS1_PRF_EMS_CHECK "tls1-prf-ems-check"
 
+/*
+ * A boolean that determines if the runtime FIPS check for undersized MAC output
+ * is performed.
+ * This is enabled by default.
```
Comment (slontis) [inline include/openssl/fips_names.h:67]: enabled??

File: providers/common/securitycheck_default.c [hunk 2]
```diff
@@ -28,6 +28,12 @@ int ossl_tls1_prf_ems_check_enabled(OSSL_LIB_CTX *libctx)
     return 0;
 }
 
+/* Disable the short MAC length check in the default provider */
+int ossl_no_short_mac_enabled(OSSL_LIB_CTX *libctx)
```
Comment (slontis) [inline providers/common/securitycheck_default.c:32]: just use the FIPS_ function and get rid of this. It has no purpose now that we have to do #ifdef FIPS_MODULE where we use it.

File: include/openssl/fips_names.h (hunk 1 above)
Comment (paulidale) [reply include/openssl/fips_names.h:67]: That's what **all** the other Booleans in the file say.

File: providers/common/securitycheck_default.c (hunk 2 above)
Comment (slontis) [reply providers/common/securitycheck_default.c:32]: At some point after everything is merged from lots of PR's we should probably change the FIPS_ name to an ossl_ name

File: providers/fips/fipsprov.c [hunk 3]
```diff
@@ -138,7 +141,7 @@ static int fips_get_params_from_core(FIPS_GLOBAL *fgbl)
     * OSSL_PROV_FIPS_PARAM_SECURITY_CHECKS and
     * OSSL_PROV_FIPS_PARAM_TLS1_PRF_EMS_CHECK are not self test parameters.
     */
-    OSSL_PARAM core_params[10], *p = core_params;
+    OSSL_PARAM core_params[12], *p = core_params;
```
Comment (slontis) [inline providers/fips/fipsprov.c:144]: This is one of the points where merging may go really bad with all the PR's

Comment (paulidale) [reply providers/fips/fipsprov.c:144]: That's why I bumped it by two -- there is a second PR that wants to modify this.
Maybe set it to something large with an assert if it overflows.

File: test/recipes/30-test_evp_data/evpmac_common.txt [hunk 4]
```diff
@@ -533,3 +533,53 @@ Input = 000102030405060708090A0B0C0D0E0F101112131415161718191A1B1C1D1E1F20212223
...
+
+Title = KMAC output is too small in FIPS
+
+FIPSversion = >=3.4.0
+MAC = KMAC256
+Key = 404142434445464748494A4B4C4D4E4F505152535455565758595A5B5C5D5E5F
+Input = 000102030405060708090A0B0C0D0E0F101112131415161718191A1B1C1D1E1F202122232425262728292A2B2C2D2E2F303132333435363738393A3B3C3D3E3F404142434445464748494A4B4C4D4E4F505152535455565758595A5B5C5D5E5F606162636465666768696A6B6C6D6E6F707172737475767778797A7B7C7D7E7F808182838485868788898A8B8C8D8E8F909192939495969798999A9B9C9D9E9FA0A1A2A3A4A5A6A7A8A9AAABACADAEAFB0B1B2B3B4B5B6B7B8B9BABBBCBDBEBFC0C1C2C3C4C5C6C7
+Output = 28c815
+Custom = "My Tagged Application"
+Unapproved = 1
```
Comment (slontis) [inline test/recipes/30-test_evp_data/evpmac_common.txt:545]: evp_test.c needs to handle this parameter

Comment (paulidale) [reply test/recipes/30-test_evp_data/evpmac_common.txt:545]: Done

**Comment by slontis** [comment]
Are the test failures valid?

File: doc/man7/EVP_MAC-KMAC.pod [hunk 5]
```diff
@@ -65,9 +64,18 @@ It is 168 for C<KMAC-128> and 136 for C<KMAC-256>.
 The "xof" parameter value is expected to be 1 or 0. Use 1 to enable XOF mode.
 The default value is 0.
 
+=item "fips-indicator" (B<OSSL_MAC_PARAM_FIPS_APPROVED_INDICATOR>) <int>
+
+This settable parameter is described in L<provider-mac(7)>.
+
+=item "no-short-mac" (B<OSSL_PROV_FIPS_PARAM_NO_SHORT_MAC>) <integer>
```
Comment (slontis) [inline doc/man7/EVP_MAC-KMAC.pod:71]: should we be more specific here about what the restriction actually is?

**Comment by paulidale** [comment]
Still digging into the test failures.  They appear to be in CMS and are failing on master.  At least mostly.

File: doc/man7/EVP_MAC-KMAC.pod (hunk 5 above)
Comment (paulidale) [reply doc/man7/EVP_MAC-KMAC.pod:71]: Added

File: doc/man7/EVP_MAC-KMAC.pod [hunk 6]
```diff
@@ -148,7 +150,8 @@ set before it instead.
 =head1 SEE ALSO
 
 L<EVP_MAC_CTX_get_params(3)>, L<EVP_MAC_CTX_set_params(3)>,
-L<EVP_MAC(3)/PARAMETERS>, L<OSSL_PARAM(3)>
+L<EVP_MAC(3)/PARAMETERS>, L<OSSL_PARAM(3)>,
+L<SP 800-185 8.4.2|https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-185.pdf>
```
Comment (slontis) [inline doc/man7/EVP_MAC-KMAC.pod:154]: Is the format ````[SP800..]https://...````

Comment (paulidale) [reply doc/man7/EVP_MAC-KMAC.pod:154]: It displays as `SP 800-185 8.4.2` and the link works, so the format is okay.

**Review by slontis — APPROVED** [review]

**Comment by slontis** [comment]
Test failures are not caused by this PR.

**Comment by slontis** [comment]
When merging any of these indicator PR's we need to very careful with anywhere where we are changing  fixed arrays sizes. Multiple PR's are all hitting these same arrays - so there WILL be merge issues where the size needs to increase,

**Comment by slontis** [comment]
(I know changing to OSSL_PARAM_BUILD might fix this, but it is too late to do that change now - and seems like overkill)

**Comment by paulidale** [comment]
ping for review @openssl/committers

File: apps/fipsinstall.c [hunk 7]
```diff
@@ -64,6 +64,7 @@ const OPTIONS fipsinstall_options[] = {
      "Forces self tests to run once on module installation"},
     {"ems_check", OPT_TLS_PRF_EMS_CHECK, '-',
      "Enable the run-time FIPS check for EMS during TLS1_PRF"},
+    {"no_short_mac", OPT_NO_SHORT_MACS, '-', "Disallow short MAC output"},
```
Comment (tom-cosgrove-arm) [inline apps/fipsinstall.c:67]: There's a slight inconsistency between `"no_short_mac"` and `OPT_NO_SHORT_MACS`

**Review by tom-cosgrove-arm — APPROVED** [review]
With other parameters we've sometimes put something in `CHANGES.md` - did you want to do that with this one?

Either way, approved

**Comment by paulidale** [comment]
Added a changes entry and resolved the naming difference.
Reconfirmation please.
Don't see a need for a timer reset.

File: CHANGES.md [hunk 8]
```diff
@@ -29,6 +29,12 @@ OpenSSL 3.4
 
 ### Changes between 3.3 and 3.4 [xx XXX xxxx]
 
+ * Add FIPS provider configuration option to enforce the a minimal
+   MAC length check.  The option '-no_short_mac' can optionally be
+   supplied to 'openssl fipsinstall'.
+
+   *Shane Lontis*
```
Comment (tom-cosgrove-arm) [inline CHANGES.md:36]: Shane or Paul?

**Review by tom-cosgrove-arm — APPROVED** [review]
Reconfirming, although not sure about the attribution in `CHANGES.md` - everything else looks good

**Comment by paulidale** [comment]
Attribution fixed and rebased to remove conflicts.

**Comment by tom-cosgrove-arm** [comment]
Hmm, CI failed at `Provider compatibility for PRs / development-branches (PR, PR.tar.gz)`

**Review by slontis — APPROVED** [review]
I would prefer to just do a second pass for the CHANGES.md after we have completed all the FIPS indicator work. It is just another point for merge conflicts.

**Comment by tom-cosgrove-arm** [comment]
@slanterns Paul has fixed `CHANGES.md` - I would re-confirm but for the failing CI - is this expected? If so, consider this a reconfirmation

**Comment by slanterns** [comment]
> @slanterns Paul has fixed `CHANGES.md` - I would re-confirm but for the failing CI - is this expected? If so, consider this a reconfirmation

Hmm, I have no idea on why do you mention me?

**Comment by paulidale** [comment]
The failing CI was a result of a different problem which has been merged now.
This results in some versioning problems appearing.

**Comment by paulidale** [comment]
Got the tests passing.  Reapprovals required (again).

File: CHANGES.md [hunk 9]
```diff
@@ -1108,7 +1114,7 @@ OpenSSL 3.1
 
  * Add FIPS provider configuration option to enforce the
    Extended Master Secret (EMS) check during the TLS1_PRF KDF.
-   The option '-ems-check' can optionally be supplied to
+   The option '-ems_check' can optionally be supplied to
```
Comment (slontis) [inline CHANGES.md:1117]: The option uses _ and  the config file uses - (hence the confusion) :)

**Review by slontis — APPROVED** [review]
I probably would have put those test changes in another PR if they have nothing to do with this PR..

**Review by tom-cosgrove-arm — APPROVED** [review]
Reconfirming.

I too would have been happy with "CI failure is unrelated" and a separate PR, but this is okay too

**Comment by paulidale** [comment]
Merged resolving trivial conflicts.
Didn't reset timer for the minor test case changes.

Thanks for the feedback.

**Pull request closed without merging**